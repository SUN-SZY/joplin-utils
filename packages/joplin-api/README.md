# joplin-api

[![npm version](https://badge.fury.io/js/joplin-api.svg)](https://www.npmjs.com/package/joplin-api)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

A JavaScript wrapper library for accessing the Joplin Data API, providing type-safe and intuitive interfaces.

## Features

- 🚀 Support multiple runtime environments: Browser, NodeJS and Joplin plugin
- 📦 Complete TypeScript type definitions
- 🔒 Type-safe API calls
- 📚 Compliant with official API specifications

## Installation

```bash
npm install joplin-api
# or
yarn add joplin-api
# or
pnpm add joplin-api
```

## Quick Start

### REST API Mode (Browser/NodeJS)

```typescript
import { joplinDataApi } from 'joplin-api'

const api = joplinDataApi({
  type: 'rest',
  token: 'your-token-here',
  baseUrl: 'http://localhost:41184',
})

// Get note list
const notes = await api.note.list()
console.log(notes)
```

### Plugin Mode (Joplin Plugin)

```typescript
import { joplinDataApi } from 'joplin-api'

const api = joplinDataApi({
  type: 'plugin',
})

// API call example
const notes = await api.note.list()
```

## Documentation

For complete API documentation, please refer to:

- [Joplin Official REST API Documentation](https://joplinapp.org/help/api/references/rest_api/)
- [Project TypeScript Type Definitions](https://joplin-utils.rxliuli.com/en-US/api/joplin-api/)

## Contributing

Issues and Pull Requests are welcome!

## License

MIT License

---

> 📝 Note: To use REST API mode, you need to enable Web Clipper service in Joplin settings and obtain an authorization token first.
