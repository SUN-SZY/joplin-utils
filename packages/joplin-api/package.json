{
  "name": "joplin-api",
  "version": "0.6.1",
  "description": "joplin ts/js data api",
  "license": "mit",
  "homepage": "https://joplin-utils.rxliuli.com/api/joplin-api",
  "repository": {
    "type": "git",
    "url": "https://github.com/rxliuli/joplin-utils/tree/master/libs/joplin-api"
  },
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "scripts": {
    "build": "vite build",
    "dev": "vite build --watch",
    "docs-build": "rimraf docs && typedoc src/index.ts",
    "docs-deploy": "pnpm docs-build && gh-pages -d docs -e /api/joplin-api --add",
    "docs-dev": "pnpm docs-build && live-server docs",
    "prepublishOnly": "pnpm build",
    "setup": "pnpm build",
    "test": "vitest run"
  },
  "dependencies": {
    "@liuli-util/object": "^3.5.1",
    "lodash-es": "^4.17.21",
    "query-string": "^9.0.0"
  },
  "devDependencies": {
    "@liuli-util/eslint-config-ts": "^0.4.0",
    "@liuli-util/vite-plugin-node": "^0.9.0",
    "@types/fs-extra": "^11.0.4",
    "@types/lodash-es": "^4.17.12",
    "@types/node": "^20.13.0",
    "fs-extra": "^11.2.0",
    "gh-pages": "^6.1.1",
    "joplin-plugin-api": "workspace:^",
    "rimraf": "^5.0.7",
    "ts-node": "^10.9.2",
    "tslib": "^2.6.2",
    "typescript": "^5.5.4",
    "vite": "^5.2.12",
    "vitest": "^1.6.0"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "deploy": {
    "type": "gh-pages",
    "dist": "docs",
    "dest": "/api/joplin-api"
  }
}
