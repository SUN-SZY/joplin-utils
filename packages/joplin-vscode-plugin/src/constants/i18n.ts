import { TranslateType } from '../i18n'
import enUS from '../i18n/en-US.json'
import zhCN from '../i18n/zh-CN.json'

export type Lang = 'en-US' | 'zh-CN'

const config = {
  language: 'en-US' as Lang,
  langs: {
    'en-US': enUS,
    'zh-CN': zhCN,
  },
}

export async function initI18n(options: { language: Lang }) {
  config.language = options.language
}

/**
 * 根据 key 获取翻译的文本
 * @param args
 */
export function t<K extends keyof TranslateType>(...args: TranslateType[K]['params']): TranslateType[K]['value'] {
  const [key, params] = args
  let result =
    (config.langs[config.language] as Record<string, string>)[key] ??
    (config.langs['en-US'] as Record<string, string>)[key]
  if (!result) {
    return key as any
  }
  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      result = result.replaceAll(`{{${key}}}`, String(value))
    })
  }
  return result as any
}
