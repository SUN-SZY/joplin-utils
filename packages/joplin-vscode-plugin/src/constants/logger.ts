import path from 'path'
import * as winston from 'winston'

export const logger = winston.createLogger({
  level: 'info',
  format: winston.format.json(),
  transports: [new winston.transports.Console({ format: winston.format.json() })],
})

export function initLogger(logPath: string) {
  logger
    .add(new winston.transports.File({ filename: path.resolve(logPath, 'error.log'), level: 'error' }))
    .add(new winston.transports.File({ filename: path.resolve(logPath, 'combined.log'), level: 'info' }))
}
