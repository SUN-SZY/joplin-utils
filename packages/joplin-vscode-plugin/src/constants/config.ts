import { ApiConfig, joplinDataApi } from 'joplin-api'
import * as vscode from 'vscode'
import { GlobalContext } from './context'

export enum SortNotesTypeEnum {
  Alphabetical = 'alphabetical',
  Default = 'default',
}

export enum SortOrderEnum {
  Asc = 'asc',
  Desc = 'desc',
}

export interface ExtConfig {
  token?: string
  baseUrl: string
  deleteConfirm?: boolean
  sortNotes?: boolean
  sortNotesType?: SortNotesTypeEnum
  sortOrder?: SortOrderEnum
}

function initApi() {
  const extConfig = vscode.workspace.getConfiguration('joplin') as vscode.WorkspaceConfiguration & ExtConfig
  GlobalContext.api = joplinDataApi({
    type: 'rest',
    baseUrl: extConfig.baseUrl,
    token: extConfig.token!,
  })
}

export async function initConfig() {
  initApi()
  vscode.workspace.onDidChangeConfiguration((ev) => {
    if (ev.affectsConfiguration('joplin.baseUrl')) {
      initApi()
    }
    if (ev.affectsConfiguration('joplin.token')) {
      initApi()
    }
  })
}
