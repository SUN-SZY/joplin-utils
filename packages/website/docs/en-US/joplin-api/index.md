# Joplin API

## Introduction

JavaScript wrapper for the Joplin data API, written in TypeScript, providing complete type definitions, including all currently publicly documented APIs.

## Usage

```ts
import { joplinDataApi } from 'joplin-api'

const api = joplinDataApi({
  type: 'rest',
  baseUrl: 'http://127.0.0.1:27583',
  token: '***',
})
// Or in plugin
const api = joplinDataApi({
  type: 'plugin',
})

const res = await api.note.list()
console.log(res)
```

> For more examples, refer to: <https://github.com/rxliuli/joplin-utils/tree/master/packages/joplin-api/src/api/__tests__>

## API Reference

| Object               | Description                                                            |
| -------------------- | ---------------------------------------------------------------------- |
| `api.event`          | Event-related API, e.g., get note modification history                 |
| `api.folder`         | Folder-related API, e.g., obtain folder tree                           |
| `api.folderExt`      | Extended folder API, e.g., move folder                                 |
| `api.joplin`         | Basic joplin API, e.g., check if joplin web service is open            |
| `api.noteAction`     | Note action API, e.g., open note in external editor                    |
| `api.note`           | Note-related API, e.g., get note content                               |
| `api.noteExt`        | Extended note API, e.g., rename                                        |
| `api.resourceAction` | Resource action API, e.g., open attachment resource in external editor |
| `api.resource`       | Resource-related API, e.g., upload image                               |
| `api.search`         | Search-related API                                                     |
| `api.tag`            | Tag-related API, e.g., modify note tags                                |
| `PageUtil`           | Static pagination utility class, e.g., get the full list of notes      |

> [API Documentation](https://joplin-utils.rxliuli.com/api/joplin-api/)
