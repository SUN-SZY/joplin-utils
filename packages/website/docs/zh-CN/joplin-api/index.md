# Joplin API

## 介绍

Joplin data API 的 JavaScript 封装，使用 TypeScript 编写，提供完整的类型定义，包含目前所有在文档中公开的 api。

## 使用

```ts
import { joplinDataApi } from 'joplin-api'

const api = joplinDataApi({
  type: 'rest',
  baseUrl: 'http://127.0.0.1:27583',
  token: '***',
})
// or in plugin
const api = joplinDataApi({
  type: 'plugin',
})

const res = await api.note.list()
console.log(res)
```

> 更多示例参考: <https://github.com/rxliuli/joplin-utils/tree/master/packages/joplin-api/src/api/__tests__>

## API 参考

| 对象                 | 描述                                              |
| -------------------- | ------------------------------------------------- |
| `api.event`          | 事件相关的 api，例如获取笔记修改历史              |
| `api.folder`         | 目录相关 api，例如获取目录树                      |
| `api.folderExt`      | 目录扩展 api，例如移动目录                        |
| `api.joplin`         | joplin 基础 api，例如检查 joplin web 服务是否打开 |
| `api.noteAction`     | 笔记相关动作 api，例如在外部编辑器打开笔记        |
| `api.note`           | 笔记相关 api，例如获取笔记的内容                  |
| `api.noteExt`        | 笔记扩展 api，例如重命名                          |
| `api.resourceAction` | 资源动作 api，例如在外部编辑器打开附件资源        |
| `api.resource`       | 资源相关 api，例如上传图片                        |
| `api.search`         | 搜索相关 api                                      |
| `api.tag`            | 标签相关 api，例如修改笔记的标签                  |
| `PageUtil`           | 分页相关的静态工具类，例如获取全量笔记列表        |

> [API 文档](https://joplin-utils.rxliuli.com/api/joplin-api/)
