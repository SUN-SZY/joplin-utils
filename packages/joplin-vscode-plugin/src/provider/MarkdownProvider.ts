import * as vscode from 'vscode'
import { TypeEnum } from 'joplin-api'
import { GlobalContext } from '../constants/context'
import { wrapLink } from '../util/markdown'
import { fromMarkdown } from 'mdast-util-from-markdown'
import { selectAll } from 'unist-util-select'
import { toString } from 'mdast-util-to-string'
import { Image, Link } from 'mdast'

type Position = {
  line: number
  column: number
}

type Range = {
  start: Position
  end: Position
}

export class MarkdownProvider implements vscode.DocumentLinkProvider {
  async resolveDocumentLink(link: vscode.DocumentLink, token: vscode.CancellationToken): Promise<vscode.DocumentLink> {
    return link
  }
  async resolveLinks(s: string): Promise<
    {
      range: Range
      id: string
      type: TypeEnum.Note | TypeEnum.Resource
      label?: string
    }[]
  > {
    const root = fromMarkdown(s)
    const links = (selectAll('link[url^=":/"],image[url^=":/"]', root) as (Link | Image)[]).filter((it) => it.url)
    const resourceIdList = new Set([...GlobalContext.openNoteResourceMap.values()].flat().map((item) => item.id))
    return links.map((it) => {
      const id = it.url.slice(2, 34)
      const label = it.type === 'link' ? toString(it.children) : it.alt ?? ''
      return {
        id,
        type: resourceIdList.has(id) || it.type !== 'link' ? TypeEnum.Resource : TypeEnum.Note,
        range: {
          start: { line: it.position!.start.line - 1, column: it.position!.start.column + label.length + 2 },
          end: { line: it.position!.end.line - 1, column: it.position!.end.column - 2 },
        },
        label,
      }
    })
  }
  async provideDocumentLinks(document: vscode.TextDocument) {
    if (!GlobalContext.openNoteMap.has(document.fileName)) {
      return
    }
    const links = await this.resolveLinks(document.getText())
    return links.map((it) => {
      const docLink = new vscode.DocumentLink(
        new vscode.Range(it.range.start.line, it.range.start.column, it.range.end.line, it.range.end.column),
        vscode.Uri.parse(wrapLink(it.id, it.type)),
      )
      docLink.tooltip = it.label ?? 'Follow link'
      return docLink
    })
  }
}
