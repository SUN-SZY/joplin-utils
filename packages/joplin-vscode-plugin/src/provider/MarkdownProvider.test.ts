import { beforeEach, expect, it } from 'vitest'
import { MarkdownProvider } from './MarkdownProvider'
import { GlobalContext } from '../constants/context'
import { pick } from 'lodash-es'
import { TypeEnum } from 'joplin-api'

let provider: MarkdownProvider
beforeEach(() => {
  provider = new MarkdownProvider()
})

it('MarkdownProvider', async () => {
  const md = `
# Some note

This link to [Another note](:/76cc95540bc54cdd9e425898e6fd25df) works great!
And this link to [Another note](:/76cc95540bc54cdd9e425898e6fd25df) causes an error because the line contains (text in brackets).
And this link to [Another note](:/76cc95540bc54cdd9e425898e6fd25df) causes an error because the line contains smile ;-)
`.trim()
  const r = await provider.resolveLinks(md)
  expect(r).matchSnapshot()
})

it('include image', async () => {
  const md = `
![test.drawio.svg](:/8c08a859711c4cfd846bf916b540ee00)
`.trim()
  const r = await provider.resolveLinks(md)
  expect(r[0].id).eq('8c08a859711c4cfd846bf916b540ee00')
  expect(r).matchSnapshot()
})

it('include resources', async () => {
  const md = `
[test.user.js](:/76cc95540bc54cdd9e425898e6fd25df)
`.trim()
  GlobalContext.openNoteResourceMap.set('76cc95540bc54cdd9e425898e6fd25df', [
    {
      id: '76cc95540bc54cdd9e425898e6fd25df',
      title: 'test.user.js',
      type_: TypeEnum.Resource,
    },
  ])
  const r = await provider.resolveLinks(md)
  expect(pick(r[0], 'id', 'type')).toEqual({
    id: '76cc95540bc54cdd9e425898e6fd25df',
    type: TypeEnum.Resource,
  })
})
