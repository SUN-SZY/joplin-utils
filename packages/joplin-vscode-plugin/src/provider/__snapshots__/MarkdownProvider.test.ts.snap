// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarkdownProvider 1`] = `
[
  {
    "id": "76cc95540bc54cdd9e425898e6fd25df",
    "label": "Another note",
    "range": {
      "end": {
        "column": 62,
        "line": 2,
      },
      "start": {
        "column": 28,
        "line": 2,
      },
    },
    "type": 1,
  },
  {
    "id": "76cc95540bc54cdd9e425898e6fd25df",
    "label": "Another note",
    "range": {
      "end": {
        "column": 66,
        "line": 3,
      },
      "start": {
        "column": 32,
        "line": 3,
      },
    },
    "type": 1,
  },
  {
    "id": "76cc95540bc54cdd9e425898e6fd25df",
    "label": "Another note",
    "range": {
      "end": {
        "column": 66,
        "line": 4,
      },
      "start": {
        "column": 32,
        "line": 4,
      },
    },
    "type": 1,
  },
]
`;

exports[`include image 1`] = `
[
  {
    "id": "8c08a859711c4cfd846bf916b540ee00",
    "label": "test.drawio.svg",
    "range": {
      "end": {
        "column": 53,
        "line": 0,
      },
      "start": {
        "column": 18,
        "line": 0,
      },
    },
    "type": 4,
  },
]
`;

exports[`include resources 1`] = `
[
  {
    "id": "8c08a859711c4cfd846bf916b540ee00",
    "label": "test.drawio.svg",
    "range": {
      "end": {
        "column": 53,
        "line": 0,
      },
      "start": {
        "column": 18,
        "line": 0,
      },
    },
    "type": 4,
  },
]
`;
